### DISCUSSION TITLE
Problem with calcParentFromVote - Variant Graph Approach

### DISCUSSION THREAD

**Discussion by Antonio-Silveira** [mtex-toolbox/mtex#1411] [Ask Anything]
Greetings everyone,

I am playing with the variant graph approach. Its really awesome!
When I use it with a kind of "bad EBSD data set", Mtex complains when I run the calcParentFromVote at the last segment:
```matlab

%% Variant Graph base parent grain reconstruction
job.calcVariantGraph('threshold',2.5*degree,'tolerance',2.5*degree)

%% Cluster the variant graph into components
job.clusterVariantGraph('includeSimilar')
%% plot best probability
plot(job.grains,job.votes.prob(:,1))
mtexColorbar

%% reconstruc all parent orientations with probability higher than 50%
job.calcParentFromVote('minProb',0.5)
% plot the Result
plot(job.parentGrains,job.parentGrains.meanOrientation)

%% Remove badly reconstructed clusters
% In order to reconstruct the remaining parent grains, we can calculate the
% votes for surrounding parent grains by the already reconstructed parent
% grains

% compute the votes
job.calcGBVotes('p2c','reconsiderAll')
% assign parent orientations according to the votes
job.calcParentFromVote    %Error happens here
% plot the result
plot(job.parentGrains,job.parentGrains.meanOrientation)
```
When it runs calcParentFromVote, the following error appears but I don't really understand why:
```matlab
Error using any
Dimension argument must be a positive integer
scalar, a vector of unique positive integers, or
'all'.

Error in subsind (line 61)
    sub = any(subs{i},
    find(size(subs{i}')==max(size(ind)),1));

Error in grain2d/subsref (line 25)
    ind = subsind(grains,s(1).subs);

Error in
parentGrainReconstructor/calcParentFromVote (line
75)
  pOri = variants(job.p2c,
  job.grainsPrior(doTransform).meanOrientation,
  job.votes.parentId(doTransform,1));

Error in Austenite_grain_reconstruction (line 95)
job.calcParentFromVote
```
To give a little bit of context, when I say "bad EBSD data set" I mean data with not the best IQ map and a bit high disorientation (~4°), leading to a sad blue probability map. I tried to modify the Threshold and Tolerance values to 4° and 4°, respectively, as Niessen says in his video, to improve the probability map, but the same error appears when I try to run this segment above. 
**Disorientation**:
![grafik](https://user-images.githubusercontent.com/103099688/165037941-5a6fdf8d-8672-4a42-99f7-5277dc3dbb6f.png)
**Probability maps**:
![grafik](https://user-images.githubusercontent.com/103099688/165039779-6e8af421-3072-45de-b5ed-2c1fe2798187.png)

Could this error mean that when I try to reconstruct the remaining parent grain the probability for the surrounding grain is 0, thus, resulting in this error? If so, besides maybe doing a better quality EBSD, is there anything else I could do to overcome this error?

Thanks in advance for any help!
Antonio

**Comment by nyyssont** [comment]
Interesting... for some reason the algorithm seems to think it needs to re-index an orientation with an existing prior orientation?

To get more information, could you try, after `job.clusterVariantGraph('includeSimilar')`:

```matlab
    doTransform = job.votes.parentId(:,1)>0 & ...
      job.votes.prob(:,1) > 0;
```

This is now a list of the grains to be transformed.

You could then try:
```matlab
pOri = variants(job.p2c, job.grainsPrior(doTransform).meanOrientation, job.votes.parentId(doTransform,1));
```

or

```matlab
  pOri = variants(job.p2c, job.grains(doTransform).meanOrientation, job.votes.parentId(doTransform,1));
```

Can you reproduce the error? One thing to try would be to run `job.clusterVariantGraph` without the option `'includeSimilar'`.

**Comment by Antonio-Silveira** [comment]
Hi Nyyssönen,

thanks for your tips.
I tried to use all the commands that you said, but nothing changed, the error persisted. However, I realized that I was using ```calcGrains``` with ebsd alone, as below: 
```matlab
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',3*degree);
% remove small grains
ebsd(grains(grains.grainSize < 5)) = [];
% reidentify grains with small grains removed:
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',3*degree);
grains = smooth(grains,5);
```
If I use exactly as written in the tutorial, then the program does not complain:
```matlab
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
```

The reason for me to use ```calcGrains(ebsd,'angle',3*degree);``` , is that I have inclusion in the microstructure, and when I use ```calcGrains(ebsd('indexed'),'angle',3*degree)``` some grains are generated inside the inclusion like the ones highlighted in blue in image(a). And if I used ``` calcGrains(ebsd('indexed'),'unitCell','angle',3*degree)  ``` I get these weird square grains in image(b). 
![grafik](https://user-images.githubusercontent.com/103099688/165313893-fa8816a2-686b-416b-90be-4dd5ea1d0bbe.png)

Is there a way to not let the grains grow inside the inclusions while still using 'indexed' so the reconstruction does not complain about it?

**Reply by Tijmenvermeij** [reply]
Might be good to check this page in the MTex documentation, there might be a good approach there.
https://mtex-toolbox.github.io/GrainReconstructionAdvanced.html
